rfd/0079/project-service-draft.md

```diff
@@ -0,0 +1,121 @@
+
+In addition to this, users should be able to associate custom tags with any given sha1 sum. 
+
+It will be easily discernible which project tags a given service changed in. Services will support custom user tags that will be associated with the service's project sha1 sum. 
+
+## Endpoints
+
+Unless otherwise noted, the output for all endpoints is JSON. The exception to this, as documented below, is for the project manifest and metadata endpoints, which will output YAML as it was provided. If the project was created via the JSON interface, then YAML will be generated. 
+
+All below endpoints may be prefixed with `/users/$userId` to access data for a user other than the one you're authenticated as, assuming authorization is granted. For example:
+
+`GET /users/abc123/projects` will return the same data as a request to `/projects` would if you were authenticated as `abc123`. 
+
+### /projects
+
+Endpoint operations:
+
+* GET - A list of projects associated with requesting user
+* POST - Create a new project via YAML manifest transmitted in POST payload
+
+### /projects/$projectId
+
+Endpoint operations:
+
+* GET - Project data and metadata derived from manifest and meta file in JSON format for specified project ID
+* PUT - Update project data and/or metadata via JSON document transmitted in PUT payload
+* DELETE - Mark project as removed
+
+### /projects/$projectId/manifest
+
+Endpoint operations:
+
+* GET - Get project manifest in YAML format as it was provided
+* PUT - Update project manifest by sending YAML payload
+
+### /projects/$projectId/metadata
+
+Endpoint operations:
+
+* GET - Get project metadata in YAML format as it was provided
+* PUT - Update project metadata by sending YAML payload
+
+### /projects/$projectId/metadata/$key
+
+Endpoint operations:
+
+* GET - Returns value associated with metadata key
+* PUT - Update value for metadata key
+* DELETE - Remove metadata key
+
+### /projects/$projectId/tags
+
+Endpoint operations:
+
+* GET - list of tags associated with project, automatic and custom
+* POST - Create a user-defined tag, associated with a specified sha1sum tag
+
+### /projects/$projectId/tags/$tagId
+
+Endpoint operations
+
+* GET - Detailed project information and manifest for the version associated with the given tag ID
+* DELETE - Remove a user-specified tag, this will return an error if user attempts to delete a sha1 sum tag
+
+### /projects/$projctId/services
+
+Endpoint operations:
+
+* GET - Summarized service information for specified project
+* POST - Create a new service for the specified project via service manifest transmitted in POST payload
+
+### /projects/$projectId/state
+
+Endpoint operations:
+
+* GET - Goal/Actual state information via a proxied request to the Convergence service
+
+### /services
+
+Endpoint operations:
+
+* GET - A list of all user services across all projects
```

Comment:
The use case I had in mind here, is I want to see a list of all my services running in production along with their associated project name (of which I have permissions for as you pointed out). It allows me to approach finding something in reverse, drilling up instead of down. Maybe I want to see all my nginx services by project, or how many projects I'm running Redis in, etc...

Getting rid of this capability would certainly simplify the endpoints. We could get rid of /services (and endpoint aliases) and just make a nice clean /project/$projectId/services/$serviceId hierarchy, but at the cost of some flexibility. Of course this could be implemented by making multiple calls to each project, but given multiple small projects, the capability of making a single call would be much faster, and thus a better user experience.